crates/sqlbuiltins/src/functions/scalars/hashing.rs

```diff
@@ -0,0 +1,96 @@
+use std::hash::{Hash, Hasher};
+
+use fnv::FnvHasher;
+use siphasher::sip::SipHasher24;
+
+use super::*;
+
+pub struct SipHash;
+
+impl ConstBuiltinFunction for SipHash {
+    const NAME: &'static str = "siphash";
+    const DESCRIPTION: &'static str =
+        "Calculates a 64bit non-cryptographic hash (SipHash24) of the value.";
+    const EXAMPLE: &'static str = "siphash(<value>)";
+    const FUNCTION_TYPE: FunctionType = FunctionType::Scalar;
+
+    fn signature(&self) -> Option<Signature> {
+        Some(Signature::new(
+            // args: <FIELD>
+            TypeSignature::Any(1),
+            Volatility::Immutable,
+        ))
+    }
+}
+impl BuiltinScalarUDF for SipHash {
+    fn as_expr(&self, args: Vec<Expr>) -> Expr {
+        let udf = ScalarUDF {
+            name: Self::NAME.to_string(),
+            signature: ConstBuiltinFunction::signature(self).unwrap(),
+            return_type: Arc::new(|_| Ok(Arc::new(DataType::Utf8))),
+            fun: Arc::new(move |input| match get_nth_scalar_value(input, 0) {
```

Comment:
I don't think this does what we want

```rust
fn get_nth_scalar_value(input: &[ColumnarValue], n: usize) -> Option<ScalarValue> {
    match input.get(n) {
        Some(input) => match input {
            ColumnarValue::Scalar(scalar) => Some(scalar.clone()),
            ColumnarValue::Array(arr) => ScalarValue::try_from_array(arr, 0).ok(),
        },
        None => None,
    }
}
```

If it's an array, it'll just use the first element. So this will likely produce invalid results when used on a column. 

sorry for being unclear in this comment. 
> can you add some slts of testing this on a column with multiple values.

What I meant to say was run the hashing function on a column with multiple **rows** of different values. I expect that they will all have the hash of the first value.

Reply:
> If it's an array, it'll just use the first element. So this will likely produce invalid results when used on a column.

The [docs here](https://arrow.apache.org/datafusion/library-user-guide/adding-udfs.html#adding-user-defined-functions-scalar-window-aggregate-table-functions) make it seem pretty clearly that scalar functions are only ever get one row [which is backed up by the example](https://github.com/apache/arrow-datafusion/blob/main/datafusion-examples/examples/simple_udf.rs) which wouldn't make sense if column references for scalar functions worked the way you're suggesting. (and the get_nth_value function [has been around for a while](https://github.com/GlareDB/glaredb/commit/d43669cb7ab49fe5af1b6eb3da80a938830ddc00#diff-1b7ec8a265c3a8ec87f20cafa22e741e7f658cbbc3c6883748d70bd14f65913dR261), so we've been doing this for a while, with no ill effects.

Reply:
in the datafusion example, you can see that they are handling the array directly. https://github.com/apache/arrow-datafusion/blob/main/datafusion-examples/examples/simple_udf.rs#L80


The `get_nth_scalar_value` function was designed to unwrap **scalars** out of ColumnarValue, not unwrap all values. It doesn't properly extract the array from `ColumnarValue::Array` variant.


If we try this on an array, we can see that it is using the first value from the array, not operating on the elements in the array. 
```sql
> create table t (x int);
Table created
> insert into t values (1), (2), (3);
Inserted 3 rows
> select siphash(x) as actual, siphash(arrow_cast(1, 'Int32')) as one from t;
┌─────────────────────┬─────────────────────┐
│              actual │                 one │
│                  ── │                  ── │
│              UInt64 │              UInt64 │
╞═════════════════════╪═════════════════════╡
│ 1308865109194702177 │ 1308865109194702177 │
│ 1308865109194702177 │ 1308865109194702177 │
│ 1308865109194702177 │ 1308865109194702177 │
└─────────────────────┴─────────────────────┘
```

Reply:
Interesting... we should probably have the `get_nth_scalar` function (and the friends I've written recently) either error for array values, or take a closure. 

Relatedly, this means that the `kdl` functions and maybe the postgres functions that use it (though perhaps more ambiguous) will have the same issue.

Reply:
I've added a test case for this.